tutorials/week5-unit-testing.md

```diff
@@ -0,0 +1,727 @@
+
+```ts
+  it("should invoke console.log() with the result 2 for inputs 1 and 1", () => {
+    const logSpy = jest.spyOn(window.console, "log");
+    logSpy.mockImplementation(() => {
+      // This will no longer print to console.
+    });
+
+    const result: number = calculator.add(1, 1);
+
+    expect(logSpy).toHaveBeenCalledWith("The result is: ", result);
+
+    logSpy.mockRestore();
+  });
+  ```
+
+_Warning:_ Watch out for circular dependencies in mock implementations.
+
+### Stub
+
+A stub is a special kind of mock which does not require an alternate implementation but instead returns some value that we specify. When a stub gets invoked, it does not invoke the actual function, but returns the desired value instead. The syntax is as below:
+
+```ts
+  spy.mockReturnValue(someValue);
+  ```
+
+To return a promise, we can use:
+
+ ```ts
+  spy.mockResolvedValue(someValue);
+  ```
+  This can be especially handy when stubbing Axios requests.
+
+Using a stub in our example simply prevents console.log() from being executed, since it does not return a value anyway.
+
+  ```ts
+  it("should invoke console.log() with the result 2 for inputs 1 and 1", () => {
+    const logSpy = jest.spyOn(window.console, "log");
+    logSpy.mockReturnValue();
+
+    const result: number = calculator.add(1, 1);
+
+    expect(logSpy).toHaveBeenCalledWith("The result is: ", result);
+
+    logSpy.mockRestore();
+  });
+  ```
+
+## Testing Asynchronous Code
+
+### Callbacks
+
+Callbacks are one of the most commonly used patterns for asynchronous programming in JavaScript/TypeScript. Consider the below callback function use case:
+
+```ts
+test('Check if I am a true husky', () => {
+  function callback(data) {
+    expect(data).toBe('I am from Northeastern!');
+  }
+  fetchData(callback);
+});
+```
+Here, fetchData() is a function that takes a callback and would call that callback function later in it's implementation. Now consider 2 scenarios:
+
+case 1: 
+```ts
+fetchData(callback) {
+  setTimeout(()=>{
+    callback('I am from Northeastern!');
+  }, 1000);
+}
+```
+
+case 2: 
+```ts
+fetchData(callback) {
+  setTimeout(()=>{
+    callback('I am not from Northeastern!');
+  }, 1000);
+}
+```
+
+The test case would still pass in either of the above scenarios since fetchData() is an async function, which means the program will not wait for the call to complete.
+
+The correct way to test a callback would be using the argument `done` in the test like shown in below example:
+
+```ts
+test('Check if I am a true husky', (done) => {
+  function callback(data) {
+    expect(data).toBe('I am from Northeastern!');
+    done();
+  }
+  fetchData(callback);
+});
+```
+
+Now, the program would wait for done to be invoked. This implementation would correctly test both the above scenarios.
+
+### Promise
+
+In previous tutorials, we have used Axios to make http requests which return promises. This is how we can write tests for axios requests. Consider the example below:
```

Comment:
Can you put this above callbacks, because we will be encouraging async programming with promises instead of callbacks?

Reply:
I have fixed this now and the `Promise` section is now above the `Callback` section.